ibis/expr/types/joins.py

```diff
@@ -236,41 +236,58 @@ def asof_join(  # noqa: D102
     ):
         predicates = util.promote_list(predicates) + util.promote_list(by)
         if tolerance is not None:
-            if not isinstance(on, str):
-                raise TypeError(
-                    "tolerance can only be specified when predicates is a string"
-                )
-            # construct a predicate with two sides from the two tables
-            predicates.append(self[on] <= right[on] + tolerance)
+            if isinstance(on, str):
+                # self is always a JoinChain so reference one of the join tables
+                left_on = self.op().values[on].to_expr()
+                right_on = right[on]
+                on = left_on >= right_on
+            elif isinstance(on, Value):
+                node = on.op()
+                if not isinstance(node, ops.Binary):
+                    raise InputTypeError("`on` must be a comparison expression")
+                left_on = node.left.to_expr()
+                right_on = node.right.to_expr()
+            else:
+                raise TypeError("`on` must be a string or a ValueExpr")
+
+            joined = self.asof_join(
+                right, on=on, predicates=predicates, lname=lname, rname=rname
+            )
+            filtered = joined.filter(
+                left_on <= right_on + tolerance, left_on >= right_on - tolerance
+            )
+            right_on = right_on.op().replace({right.op(): filtered.op()}).to_expr()
```

Comment:
Dereferencing fails to resolve it, manual replacement works. It would be nice to have the dereferencing mechanism support this case.